### ISSUE TITLE
idea: treat a major coverage drop an issue!

### ISSUE THREAD

**Issue by bagder** [google/oss-fuzz#11398]
The curl code base has been fuzzed by oss-fuzz for many years. These days we rarely get any reports. It chugs away in the background and while it certainly gives us confidence and helps us "stay honest", it also falls off the top of our minds. We don't think about it too much in our daily work. I think that's good. It still tells us when it finds something. Which these days are many months between each time.

With this in mind, when we changed a config detail **a year ago**, the fuzz coverage for curl dropped **considerably** but no one noticed because no one checked. We are used to oss-fuzz telling us when there's an issue. Thus, oss-fuzz has the last year been fuzzing a much smaller part of the code base, which then of course could have made us miss problems.

The other day we fixed this issue and the coverage is again back up to levels similar to those a year ago, and nothing new has been reported (yet) so I figure in a way we were lucky and dodged one this time. But it brings up this topic on my agenda: **how do I detect this problem the next time**?

And perhaps even more broadly: how do we all (all oss-fuzz users) detect this because I presume that even though it happened to us this time, it might happen to others every once in a while. And maybe it has already happened to 22 users who have not even noticed yet and now the fuzzer is missing out?

How do others do? Is there a best practice I have missed? Can oss-fuzz perhaps itself detect huge drops and file an issue about that?

**Comment by evverx** [comment]
> How do others do? Is there a best practice I have missed?

I opened https://github.com/google/oss-fuzz/issues/7190 so that CIFuzz could coverage drops as soon as possible but it kind of stalled. The only way to keep track of drops like that I'm aware of is the Fuzz-Introspector dashboard but it works when there are less than ten fuzz targets. With more than 40 fuzz targets it gets blurry.

**Comment by DavidKorczynski** [comment]
> but it works when there are less than ten fuzz targets.

Could you clarify this? It works when there are 10 or more as well?

FWIF it's visible that there was a sudden jump in Curl's coverage: https://introspector.oss-fuzz.com/project-profile?project=curl but I think the drop may have happened at a point before the app was started (April 20 2023)

**Comment by evverx** [comment]
> Could you clarify this? It works when there are 10 or more as well?

The global dashboard itself works with a lot of fuzz targets. The problem is that sometimes coverage drops affect certain fuzz targets and it's hard to spot them because they are hidden behind the global coverage where several fuzz targets can cover the same code a bit differently. Basically I have to go over specific coverage reports one by one.

**Comment by phi-go** [comment]
FYI this feature now exists: https://github.com/ossf/fuzz-introspector/pull/2102

I would wait for first results but we could maybe close this issue then.

@bagder @evverx to gauge if this should be implemented, would you want to be notified if there are issues detected for a fuzz target or is the project view enough?

**Comment by evverx** [comment]
> would you want to be notified

I think it depends. If the coverage goes down significantly I think it should be certainly reported but ideally thresholds and things like that should be configurable probably to avoid spurious notifications.

The latest coverage drop was https://github.com/google/oss-fuzz/issues/12519#issuecomment-2558144965 (where one fuzz target went down) and as far as I understand with https://github.com/ossf/fuzz-introspector/pull/2102 it should be easier to spot things like that.

**Comment by phi-go** [comment]
So would you prefer to be notified by default, let's say if there is a 5% drop in a fuzzer, or only after configuration for a threshold has been set?

**Comment by evverx** [comment]
I think there are two (kind of) different use cases here. It's probably safe to say that all the projects should be notified by default when their **total** coverage goes down significantly. Drops affecting specific fuzz targets/code paths are important too (for example https://github.com/bytecodealliance/wasmtime/security/advisories/GHSA-5fhj-g3p3-pq9g) but it's not always obvious how to deal with them. Personally I'd look at the dashboards every now and then for the time being to see how it works first.

**Comment by phi-go** [comment]
Thank you for the input! So to put this into actionable points:

- There should be opt-out warnings for significant coverage drops of total project coverage.
- Have opt-out warnings for the case when a fuzz target ceases to built successfully or if a target starts to be blocked.
- Have opt-in warnings for significant fuzz target specific coverage drops.
- Treat a 5% coverage drop or a 1000 line drop in coverage as significant but make these values configurable.
- Reassess once this feature has been around for some time.

What do you think?

**Comment by evverx** [comment]
> There should be opt-out warnings for significant coverage drops of total project coverage.
> Have opt-out warnings for the case when a fuzz target ceases to built successfully or if a target starts to be blocked.

@phi-go I'm a bit reluctant when it comes to notifications because I'm subscribed to a bunch of different projects with different levels of tolerance to false positives so personally I think before adding opt-out notifications it would be great to figure out what it should do when fuzz targets are renamed, new code is added and so on. I don't think those things shouldn't trigger notifications to, say, mailing lists with hundreds of people there.

**Comment by phi-go** [comment]
Yeah, that's why I'm asking. I would expect the reports to be sent to the issue tracker as is already done, are these monitored by large mailing lists? But yes, it is probably better to keep everything opt-in at the beginning, maybe forever.

From my experience renaming fuzz targets does not happen often. This could maybe be detected by code matching, though, this seems a bit error prone and difficult to implement.

Coverage drops due to adding code happens often (obviously). But is this truly a false positive if the fuzzing setup has reduced coverage due to added project code? In a perfect world harnessing should be updated to address new code so a notification seems warranted. Adding code could also break the fuzzer harness. Though, I'm not so eager on having notification when vendoring third-party code, this also regularly causes coverage drops. However, here we would need to separate project code and third-party code, which is likely not trivial, probably another thing that should be configurable. (See here for a bit more discussion: https://github.com/ossf/fuzz-introspector/issues/2054)

Another [common reason](https://github.com/google/oss-fuzz/issues/12986#issuecomment-2624742095) for coverage changes commonly seen as fluctuations are that the coverage process does not run through cleanly (think OOM or timeout). Should this be reported to maintainers?

**Comment by phi-go** [comment]
> Agreed. Though I renamed fuzz targets a few times and it would have been weird if OSS-Fuzz had filed bug reports and sent notifications.

Yes, fair enough. What do you think of optionally allowing maintainers to specify a list of fuzz targets that should be fuzzed? We can use a heuristic based on fuzzers available in the past if this list is not specified. Renaming a fuzz target would still cause an alert in the later case.

Also personally I would actually like to see that the fuzz targets have changed as the oss-fuzz interaction is pretty much async this would signal to me that my change worked but I guess having a dashboard the same can be confirmed.

> In "umbrella" projects it's not always clear-cut. With no context I think it should stick to reporting obvious regressions it can be 100% sure about.

While there is the difference that for a JSON parser the new code should be covered by the existing harness this is not necessarily the case for a more diverse project. However, in both cases new code is added that should be fuzzed, is that not a clear regression? Or do you mean that the new code might not need to be fuzzed?

I'm probably missing something but without context how can we detect obvious regressions?

> I think those things should be more visible but I'm not sure they should always lead to notifications. It seems to me dashboards should be a good place to experiment, collect project-specific data and act on those things.

I agree, let's take it slow and extend the dashboard, we can still do notifications later.

**Comment by evverx** [comment]
> What do you think of optionally allowing maintainers to specify a list of fuzz targets that should be fuzzed?

It can probably work if projects are actively involved in maintaining their fuzzing infrastructure but I don't think there are a lot of projects where it happens in practice. That's why I don't think that manual tweaks can work in practice in general. There are FI issues where it's discussed how maintainers can manually specify their threat models, exclude coverage and so on but it's unlikely a lot of projects are going to do that.

> the oss-fuzz interaction is pretty much async

Agreed. That's why I think CIFuzz (which is run when PRs are opened) should do as much as possible to make sure stuff works/prevent regressions from landing. It's not there yet though.

> However, in both cases new code is added that should be fuzzed, is that not a clear regression? Or do you mean that the new code might not need to be fuzzed?

I was trying to say that in some cases it's hard to tell whether code should be fuzzed at all. It's also worth mentioning that things like static reachability and so on aren't available everywhere because there are projects where FI just doesn't work.

> I'm probably missing something but without context how can we detect obvious regressions?

As far as I understand this issue was prompted by a huge coverage drop in the curl project

![Image](https://github.com/user-attachments/assets/7afe9e4f-a81b-418f-bc25-fe3dad8792fa)

and I think in cases like this it's enough to warrant a bug report with almost no context. I think there are things that need looking into (like how to auto-close those issues when nobody reacts to them for a year and a lot of things have changed in the projects) but I think there should be a way for projects to say that they are interested in monitoring that metric.

**Comment by phi-go** [comment]
> Agreed. That's why I think CIFuzz (which is run when PRs are opened) should do as much as possible to make sure stuff works/prevent regressions from landing. It's not there yet though.

I haven't used CIFuzz and only took a short look, is there a dashboard available that shows fuzzing results? Or how are the results for CIFuzzing currently made available. I saw that they are included in a SARIF report for systemd is that usual?

> As far as I understand this issue was prompted by a huge coverage drop in the curl project

That is my understanding as well, however, huge coverage drops are actually surprisingly common due to vendoring third party code and adding project code, less often this happens because of partial build failures which I think was the case for curl. However, I guess this can be accounted for by looking at code size increases as well.

**Comment by evverx** [comment]
> is there a dashboard available that shows fuzzing results? Or how are the results for CIFuzzing currently made available. I saw that they are included in a SARIF report for systemd is that usual?

There was a global CIFuzz dashboard but I don't think it got past the proof-of-concept stage. The results are reported in PRs (by analogy with all the other GHActions) and the SARIF reports were added to make it easier to spot crashes (in the sense that instead of looking for red crosses among a gazillion CI jobs and looking at the logs it should be possible to just wait for comments based on backtraces). It's now included in the documentation so projects copying the example should get it working. Projects where CIFuzz was integrated earlier don't necessarily have it enabled though.

> huge coverage drops are actually surprisingly common due to vendoring third party code and adding project code

I took a quick look at https://github.com/ossf/fuzz-introspector/issues/2054 and as far as I can see it's discussed there. That being said `grpc-httpjson-transcoding` is used as an example there and their graph is relatively flat:

![Image](https://github.com/user-attachments/assets/e369d82f-ef9f-495c-9b37-c63819e0210c)

Could it be that different metrics are discussed as coverage drops?

**Comment by evverx** [comment]
> because of partial build failures which I think was the case for curl

Just for completeness sake
https://ostif.org/wp-content/uploads/2024/02/cURL-HTTP3-Comprehensive-Report.pdf

> this occurred after an
OpenSSL version upgrade. The new OpenSSL version started installing the libssl.a static
library on a different directory, lib64, instead of the traditional lib folder. The cURL fuzz
scripts did not expect nor support this alternate location and therefore built cURL without
SSL support, which broke several expectations in the fuzzing harnesses

**Comment by phi-go** [comment]
> There was a global CIFuzz dashboard but I don't think it got past the proof-of-concept stage. The results are reported in PRs (by analogy with all the other GHActions) and the SARIF reports were added to make it easier to spot crashes (in the sense that instead of looking for red crosses among a gazillion CI jobs and looking at the logs it should be possible to just wait for comments based on backtraces). It's now included in the documentation so projects copying the example should get it working. Projects where CIFuzz was integrated earlier don't necessarily have it enabled though.

Interesting, thank you for the overview! Does the CIFuzz run compare against the previous run or is it "just" a fuzzing run? So could coverage drops be detected?

>  That being said grpc-httpjson-transcoding is used as an example there and their graph is relatively flat

The graph does not go back far enough, this initially happened late 2021, see below. If you are interested you can find notes I did on 5% coverage drops [here](https://anonymous.4open.science/r/fuzz-harness-degradation-submission-C1E7/case_studies/notes.md). Though, while I looked into each case a bit, this was still relatively quick and I now understand oss-fuzz / introspector better. For example, I would guess that pretty much all low queue entry issues are OOMs or timeouts during coverage measurement, statefulness could also be a reason but it seems less likely to me.

![Image](https://github.com/user-attachments/assets/aa6127de-de5d-4250-887e-bd931f7b6e8b)

**Comment by phi-go** [comment]
> Currently it builds and runs fuzz targets. To figure out what it should run it downloads the latest coverage reports and looks at what is affected by diffs. In theory it can also build coverage reports and compare them (by analogy with how coverage services like Coveralls, Codecov and so on work). The upside is that people using those services are kind of used to flakiness so it can probably be more flaky and overzealous than OSS-Fuzz itself.

I see, thanks again. Maybe I will have some time to contribute to CIFuzz.

> Got it. It's a lot of process so I just took a quick look. (Assuming it's going to be published eventually it would be great if you could share a link to the paper if possible).

Currently rewriting the paper a bit but happy to share a pre-print in a few weeks.

**Comment by evverx** [comment]
> a pre-print

Just for future reference as far as I understand it's https://arxiv.org/abs/2505.06177 (An Empirical Study of Fuzz Harness Degradation). Based on that research FI now even shows coverage drops caused by OOMs/timeouts in specific flaky fuzz targets.

I'm still not sure how coverage drops can be reported by OSS-Fuzz as issues reliably (so personally I'd opt out if that feature was available because FI covers those things without annoying notifications. The problem is that FI has to be monitored though and it's probably not very convenient for everybody).

**Comment by phi-go** [comment]
Hey, sorry I forgot to follow up on this but you found the correct pre-print. There is actually a newer version of the paper as it is now under submission again but the results pretty much stayed the same. (I still need to get around to updating it on ArXiv.)

I agree coverage drops are better monitored via FI. If the FI team decides to support notifications on coverage drops, I would exclude fuzz targets that have errors as these should be looked at anyway and the coverage results they do produce are not reliable. For other fuzz targets the coverage results are really quite stable and I think it is possible to get useful signal from that data.
